src/VernaContextProvider.tsx

```diff
@@ -0,0 +1,102 @@
+import * as React from 'react';
+import { PropsWithChildren, useState } from 'react';
+import { JSONSchema7 } from 'json-schema';
+import type { UiSchema, Widget } from '@rjsf/core';
+import _ from 'lodash';
+
+function functionNotSet() {
+  throw new Error('function context not set');
+}
+
+export interface WidgetType {
+  [name: string]: Widget;
+}
+
+interface VernaContextProps {
+  schema: JSONSchema7;
+  setSchema: (newSchema: JSONSchema7) => void;
+  uiSchema: UiSchema;
+  setUiSchema: (newUiSchema: UiSchema) => void;
+  widgets: WidgetType;
+  setWidgets: (newWidgets: WidgetType) => void;
+  readOnly: boolean;
+  setReadOnly: (newMode: boolean) => void;
+}
+
+const VernaContext = React.createContext<VernaContextProps>({
+  schema: {},
+  setSchema: () => functionNotSet(),
+  uiSchema: {},
+  setUiSchema: () => functionNotSet(),
+  widgets: {},
+  setWidgets: () => functionNotSet(),
+  readOnly: true,
+  setReadOnly: () => functionNotSet(),
+});
+
+function useVerna() {
+  const context = React.useContext(VernaContext);
+  if (!context) {
+    throw new Error('UseVerna must be used withing a VernaContextProvider');
+  }
+  return context;
+}
+
+interface VernaContextProviderProps {
+  defaultSchema?: JSONSchema7;
+  defaultUiSchema?: UiSchema;
+  defaultWidget?: WidgetType;
+  defaultReadOnly?: boolean;
+}
+
+function genDefaultUiSchema(schema: JSONSchema7, uiSchema: UiSchema) {
+  if (uiSchema && uiSchema['ui:order']) return uiSchema;
+  const defaultui: UiSchema = {
+    ...uiSchema,
+    'ui:order': _.keys(schema?.['properties']),
+  };
+  _.forEach(schema?.['properties'], (prop, key) => {
+    const schemaElement = prop as JSONSchema7;
+    if (schemaElement['properties']) {
+      defaultui[key] = {
+        ...uiSchema?.[key],
+        'ui:order': _.keys(schemaElement['properties']),
+      };
+    }
+  });
+  return defaultui;
+}
+
+function VernaContextProvider({
+  defaultSchema,
+  defaultUiSchema,
+  defaultWidget,
+  defaultReadOnly,
+  children,
+}: PropsWithChildren<VernaContextProviderProps>) {
+  const [schema, setSchema] = useState<JSONSchema7>(defaultSchema || {});
+  const [uiSchema, setUiSchema] = useState<UiSchema>(
+    genDefaultUiSchema(defaultSchema || {}, defaultUiSchema || {}),
+  );
+  const [readOnly, setReadOnly] = useState<boolean>(defaultReadOnly || false);
+  const [widgets, setWidgets] = useState<WidgetType>(defaultWidget || {});
```

Comment:
You could use `defaultProps` instead.
```suggestion
  const [schema, setSchema] = useState<JSONSchema7>(defaultSchema);
  const [uiSchema, setUiSchema] = useState<UiSchema>(
    genDefaultUiSchema(defaultSchema, defaultUiSchema),
  );
  const [readOnly, setReadOnly] = useState<boolean>(defaultReadOnly);
  const [widgets, setWidgets] = useState<WidgetType>(defaultWidget);
  
  ...
  VernaContextProvider.defaultProps = {
    defaultSchema: {},
    defaultUiSchema: {},
    defaultWidget: {},
    defaultReadOnly: false,
  }
```

Reply:
Good idea !